Gems/Terrain/Code/Source/TerrainRaycast/TerrainRaycastContext.cpp

```diff
@@ -0,0 +1,393 @@
+                                        AzFramework::RenderGeometry::RayResult& result)
+    {
+        const AZ::Vector3 edgeAB = vertexB - vertexA;
+        const AZ::Vector3 edgeAC = vertexC - vertexA;
+        const AZ::Vector3 pVec = rayDirection.Cross(edgeAC);
+        const float det = edgeAB.Dot(pVec);
+        if (AZ::IsClose(det, 0.0f))
+        {
+             // The ray is parallel to the triangle.
+            return;
+        }
+
+        const float detInv = 1.0f / det;
+        const AZ::Vector3 tVec = rayStart - vertexA;
+        const float u = detInv * tVec.Dot(pVec);
+        if (u < 0.0f || u > 1.0f)
+        {
+            // No intersection.
+            return;
+        }
+
+        const AZ::Vector3 qVec = tVec.Cross(edgeAB);
+        const float v = detInv * rayDirection.Dot(qVec);
+        if (v < 0.0 || u + v > 1.0)
+        {
+            // No intersection.
+            return;
+        }
+
+        const float t = detInv * edgeAC.Dot(qVec);
+        if (t > FLT_EPSILON)
+        {
+            result.m_worldPosition = rayStart + (rayDirection * t);
+            result.m_worldNormal = edgeAB.Cross(edgeAC);
+            result.m_distance = rayDirection.GetLength() * t;
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Convenience function to get the terrain height values at each corner of an AABB, triangulate them,
+    // and then find the nearest intersection (if any) between the resulting triangles and the given ray.
+    inline void TriangulateAndFindNearestIntersection(const TerrainSystem& terrainSystem,
+                                                      const AZ::Aabb& aabb,
+                                                      const AZ::Vector3& rayStart,
+                                                      const AZ::Vector3& rayDirection,
+                                                      const AZ::Vector3& rayDirectionReciprocal,
+                                                      AzFramework::RenderGeometry::RayResult& result)
+    {
+        // Obtain the height values at each corner of the AABB.
+        const AZ::Vector3& aabbMin = aabb.GetMin();
+        const AZ::Vector3& aabbMax = aabb.GetMax();
+        AZ::Vector3 point0 = aabbMin;
+        AZ::Vector3 point2 = aabbMax;
+        AZ::Vector3 point1(point0.GetX(), point2.GetY(), 0.0f);
+        AZ::Vector3 point3(point2.GetX(), point0.GetY(), 0.0f);
+        point0.SetZ(terrainSystem.GetHeight(point0, AzFramework::Terrain::TerrainDataRequests::Sampler::DEFAULT));
+        point1.SetZ(terrainSystem.GetHeight(point1, AzFramework::Terrain::TerrainDataRequests::Sampler::DEFAULT));
+        point2.SetZ(terrainSystem.GetHeight(point2, AzFramework::Terrain::TerrainDataRequests::Sampler::DEFAULT));
+        point3.SetZ(terrainSystem.GetHeight(point3, AzFramework::Terrain::TerrainDataRequests::Sampler::DEFAULT));
```

Comment:
Suggestion:  When optimizing, might want to measure if using ProcessHeightsFromList() provides any gains here.